File: Lectures/Lecture_12/Factorize.pl
#!/usr/bin/env perl

##  Factorize.pl
##  Author: Avi Kak
##  Date: February 19, 2016 

##  Uncomment line (F12) and comment out line (F13) if you want to see the results
##  with the simpler form of the Pollard-Rho algorithm.

use strict;
use warnings;

die "\nUsage:   $0  <integer> \n" unless @ARGV == 1;                         #(A1)
my $p = shift @ARGV;                                                         #(A2)

die "Your number is too large for factorization by this script. " .
    "Instead, try the script 'FactorizeWithBigInt.pl'\n"
    if $p > 0x7f_ff_ff_ff;                                                   #(A3)

my @factors = @{factorize($p)};                                              #(A4)
my %how_many_of_each;                                                        #(A5)
map {$how_many_of_each{$_}++} @factors;                                      #(A6)
print "\nFactors of $p:\n";                                                  #(A7)
foreach my $factor (sort {$a <=> $b} keys %how_many_of_each) {               #(A8)
    print "    $factor ^ $how_many_of_each{$factor}\n";                      #(A9)
}                                           

sub factorize {                                                              #(F1)
    my $n = shift;                                                           #(F2)
    my @prime_factors = ();                                                  #(F3)
    my @factors;                                                             #(F4)
    push @factors, $n;                                                       #(F5)
    while (@factors > 0) {                                                   #(F6)
        my $p = pop @factors;                                                #(F8)
        if (test_integer_for_prime($p)) {                                    #(F9)
            push @prime_factors, $p;                                         #(F10)
            next;                                                            #(F11)
        }
#        my $d = pollard_rho_simple($p);                                     #(F12)
        my $d = pollard_rho_strong($p);                                      #(F13)
        if ($d == $p) {                                                      #(F14)
            push @factors, $d;                                               #(F15)
        } else {                   
            push @factors, $d;                                               #(F16)       
            push @factors, int($p / $d);                                     #(F17)
        }
    }
    return \@prime_factors;                                                  #(F18)
}

sub test_integer_for_prime {                                                 #(P1)
    my $p = shift;                                                           #(P2)
    my @probes = qw[ 2 3 5 7 11 13 17 ];                                     #(P3)
    foreach my $a (@probes) {                                                #(P4)
        return 1 if $a == $p;                                                #(P5)
    }
    my ($k, $q) = (0, $p - 1);                                               #(P6)
    while (! ($q & 1)) {                                                     #(P7)
        $q >>= 1;                                                            #(P8)
        $k += 1;                                                             #(P9)
    }
    my ($a_raised_to_q, $a_raised_to_jq, $primeflag);                        #(P10)
    foreach my $a (@probes) {                                                #(P11)
        my ($base,$exponent) = ($a,$q);                                      #(P12)
        my $a_raised_to_q = 1;                                               #(P13)
        while ((int($exponent) > 0)) {                                       #(P14)
            $a_raised_to_q = ($a_raised_to_q * $base) % $p 
                                              if int($exponent) & 1;         #(P15)
            $exponent = $exponent >> 1;                                      #(P16)
            $base = ($base * $base) % $p;                                    #(P17)
        }
        next if $a_raised_to_q == 1;                                         #(P18)
        next if ($a_raised_to_q == ($p - 1)) && ($k > 0);                    #(P19)
        $a_raised_to_jq = $a_raised_to_q;                                    #(P20)
        $primeflag = 0;                                                      #(P21)
        foreach my $j (0 .. $k - 2) {                                        #(P22)
            $a_raised_to_jq = ($a_raised_to_jq ** 2) % $p;                   #(P23)
            if ($a_raised_to_jq == $p-1) {                                   #(P24)
                $primeflag = 1;                                              #(P25)
                last;                                                        #(P26)
            }
        }
        return 0 if ! $primeflag;                                            #(P27)
    }
    my $probability_of_prime = 1 - 1.0/(4 ** scalar(@probes));               #(P28)
    return $probability_of_prime;                                            #(P29)
}
    
sub pollard_rho_simple {                                                     #(Q1)
    my $p = shift;                                                           #(Q2)
    my @probes = qw[ 2 3 5 7 11 13 17 ];                                     #(Q3)
    foreach my $a (@probes) {                                                #(Q4)
        return $a if $p % $a == 0;                                           #(Q5)
    }
    my $d = 1;                                                               #(Q6)
    my $a = 2 + int(rand($p));                                               #(Q7)
    my @random_num = ($a);                                                   #(Q8)
    while ($d == 1) {                                                        #(Q9)
        my $b = 2 + int(rand($p));                                           #(Q10)
        foreach my $a (@random_num) {                                        #(Q11)
            $d = gcd($a - $b, $p);                                           #(Q12)
            last if $d > 1;                                                  #(Q13)
        }
        push @random_num, $b;                                                #(Q14)
    }
    return $d;                                                               #(Q15)
}
    
sub pollard_rho_strong {                                                     #(R1)
    my $p = shift;                                                           #(R2)
    my @probes = qw[ 2 3 5 7 11 13 17 ];                                     #(R3)    
    foreach my $a (@probes) {                                                #(R4)
        return $a if $p % $a == 0;
    }
    my $d = 1;                                                               #(R5)
    my $a = 2 + int(rand($p));                                               #(R6)
    my $c = 2 + int(rand($p));                                               #(R6)
    my $b = $a;                                                              #(R7)
    while ($d == 1) {                                                        #(R8)
        $a = ($a * $a + $c) % $p;                                            #(R9)
        $b = ($b * $b + $c) % $p;                                            #(R10)
        $b = ($b * $b + $c) % $p;                                            #(R11)
        $d = gcd($a - $b, $p);                                               #(R12)
        last if $d > 1;                                                      #(R13)
    }
    return $d;                                                               #(R14)
}

sub gcd {                                                                    #(S1)
    my ($a,$b) = @_;                                                         #(S2)
    while ($b) {                                                             #(S3)
        ($a,$b) = ($b, $a % $b);                                             #(S4)
    }
    return $a;                                                               #(S5)
}

File: Lectures/Lecture_12/Factorize.py
#!/usr/bin/env python

##  Factorize.py
##  Author: Avi Kak
##  Date: February 26, 2011 
##  Modified: Febrary 25, 2012

##  Uncomment line (F9) and comment out line (F10) if you want to see the results
##  with the simpler form of the Pollard-Rho algorithm.

import random
import sys

def factorize(n):                                                            #(F1)
    prime_factors = []                                                       #(F2)
    factors = [n]                                                            #(F3)
    while len(factors) != 0:                                                 #(F4)
        p = factors.pop()                                                    #(F5)
        if test_integer_for_prime(p):                                        #(F6)
            prime_factors.append(p)                                          #(F7)
            continue                                                         #(F8)
#        d = pollard_rho_simple(p)                                           #(F9)
        d = pollard_rho_strong(p)                                            #(F10)
        if d == p:                                                           #(F11)
            factors.append(d)                                                #(F12)
        else:                                                                #(F13)
            factors.append(d)                                                #(F14)
            factors.append(p//d)                                             #(F15)
    return prime_factors                                                     #(F16)

def test_integer_for_prime(p):                                               #(P1)
    probes = [2,3,5,7,11,13,17]                                              #(P2)
    for a in probes:                                                         #(P3)
        if a == p: return 1                                                  #(P4)
    if any([p % a == 0 for a in probes]): return 0                           #(P5)
    k, q = 0, p-1                                                            #(P6)
    while not q&1:                                                           #(P7)
        q >>= 1                                                              #(P8)
        k += 1                                                               #(P9)
    for a in probes:                                                         #(P10)
        a_raised_to_q = pow(a, q, p)                                         #(P11)
        if a_raised_to_q == 1 or a_raised_to_q == p-1: continue              #(P12)
        a_raised_to_jq = a_raised_to_q                                       #(P13)
        primeflag = 0                                                        #(P14)
        for j in range(k-1):                                                 #(P15)
            a_raised_to_jq = pow(a_raised_to_jq, 2, p)                       #(P16)
            if a_raised_to_jq == p-1:                                        #(P17)
                primeflag = 1                                                #(P18)
                break                                                        #(P19)
        if not primeflag: return 0                                           #(P20)
    probability_of_prime = 1 - 1.0/(4 ** len(probes))                        #(P21)
    return probability_of_prime                                              #(P22)

def pollard_rho_simple(p):                                                   #(Q1)
    probes = [2,3,5,7,11,13,17]                                              #(Q2)
    for a in probes:                                                         #(Q3)
        if p%a == 0: return a                                                #(Q4)
    d = 1                                                                    #(Q5)
    a = random.randint(2,p)                                                  #(Q6)
    random_num = []                                                          #(Q7)
    random_num.append( a )                                                   #(Q8)
    while d==1:                                                              #(Q9)
        b = random.randint(2,p)                                              #(Q10)
        for a in random_num[:]:                                              #(Q11)
            d = gcd( a-b, p )                                                #(Q12)
            if d > 1: break                                                  #(Q13)
        random_num.append(b)                                                 #(Q14)
    return d                                                                 #(Q15)

def pollard_rho_strong(p):                                                   #(R1)
    probes = [2,3,5,7,11,13,17]                                              #(R2)
    for a in probes:                                                         #(R3)
        if p%a == 0: return a                                                #(R4)
    d = 1                                                                    #(R5)
    a = random.randint(2,p)                                                  #(R6)
    c = random.randint(2,p)                                                  #(R7)
    b = a                                                                    #(R8)
    while d==1:                                                              #(R9)
        a = (a * a + c) % p                                                  #(R10)
        b = (b * b + c) % p                                                  #(R11)
        b = (b * b + c) % p                                                  #(R12)
        d = gcd( a-b, p)                                                     #(R13)
        if d > 1: break                                                      #(R14)
    return d                                                                 #(R15)

def gcd(a,b):                                                                #(S1)
    while b:                                                                 #(S2)
        a, b = b, a%b                                                        #(S3)
    return a                                                                 #(D4)

if __name__ == '__main__':                                                   #(A1)

    if len( sys.argv ) != 2:                                                 #(A2)
        sys.exit( "Call syntax:  Factorize.py  number" )                     #(A3)
    p = int( sys.argv[1] )                                                   #(A4)
    factors = factorize(p)                                                   #(A5)
    print("\nFactors of %d:" % p)                                            #(A6)
    for num in sorted(set(factors)):                                         #(A7)
        print("%s %d ^ %d" % ("    ", num, factors.count(num)))              #(A8)
